Can't compile example classes.rs

Hello, thanks for the cool project.

I cannot compile the classes example:
https://github.com/boa-dev/boa/blob/master/boa/examples/classes.rs

I imported boa like this: `Boa = "0.11.0"`
I get the same error on both stable and nightly.

```
$ cargo run

    Blocking waiting for file lock on build directory
   Compiling jsint v0.1.0 (/tmp/jsint)
error[E0433]: failed to resolve: could not find `gc` in the list of imported crates
  --> src/main.rs:15:17
   |
15 | #[derive(Debug, Trace, Finalize)]
   |                 ^^^^^ could not find `gc` in the list of imported crates
   |
   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `gc` in the list of imported crates
  --> src/main.rs:15:24
   |
15 | #[derive(Debug, Trace, Finalize)]
   |                        ^^^^^^^^ could not find `gc` in the list of imported crates
   |
   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `gc` in the list of imported crates
  --> src/main.rs:15:17
   |
15 | #[derive(Debug, Trace, Finalize)]
   |                 ^^^^^ not found in `gc`
   |
   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider importing one of these items
   |
1  | use boa::gc::Trace;
   |
1  | use crate::Trace;
   |

error[E0433]: failed to resolve: could not find `gc` in the list of imported crates
  --> src/main.rs:15:17
   |
15 | #[derive(Debug, Trace, Finalize)]
   |                 ^^^^^ not found in `gc`
   |
   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider importing one of these items
   |
1  | use boa::gc::Finalize;
   |
1  | use crate::Finalize;
   |

error: aborting due to 4 previous errors

```


I get the same error every time I try to derive Trace or Finalize.
I tested with both the stable and nightly toolchain.

- OS: debian 
- Version: testing
- Target triple: Debian 5.10.24-1 (2021-03-19) x86_64 GNU/Linux
- Rustc version: rustc 1.54.0-nightly (5c0292654 2021-05-11)
- Stable rust: rustc 1.52.1 (9bc8c42bb 2021-05-09)

I also ran into this issue while trying to embed Boa into my app. A few key issues stand out to me here:

(note: I am running this code in my own testing project; my `main.rs` is identical to Boa's `boa/examples/classes.rs`)

1. Your issue can be resolved by importing `gc` into your `Cargo.toml` (I used version 0.4)
2. Upon applying that change, the example shows a __new__ error regarding the derive macro for `boa::gc::Trace`:
```sh
$ cargo check
...
error[E0277]: the trait bound `Person: boa::gc::Trace` is not satisfied
  --> src/main.rs:49:6
   |
49 | impl Class for Person {
   |      ^^^^^ the trait `boa::gc::Trace` is not implemented for `Person`
   | 
  ::: /home/carter/.cargo/registry/src/github.com-1ecc6299db9ec823/Boa-0.11.0/src/class.rs:71:18
   |
71 | pub trait Class: NativeObject + Sized {
   |                  ------------ required by this bound in `boa::class::Class`

error: aborting due to previous error
```
3. Assuming that the `console` feature is enabled, this example will compile fine using the upstream git link rather than version 0.11:
```toml
Boa = {git = "https://github.com/boa-dev/boa", features = ["console"]}
```
4. However, this will result in an Error unwrap after `eval()`, that stems from the method `person.sayHello()`. Removing the use of this method from the JS source code will completely evaluate the code without errors.

[(Output of the previous step)](https://pastebin.com/1QvUCL7x)

- System: Linux 5.12.12-arch1-1, Arch Linux + KDE Plasma 5.22.1
- Target triple: x86_64-unknown-linux-gnu
- rustc 1.55.0-nightly (6a758ea7e 2021-06-22)

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -212,10 +212,16 @@ jobs:
             ~/.cargo/git
             ~/.cargo/registry
           key: ${{ runner.os }}-cargo-examples-${{ hashFiles('**/Cargo.lock') }}
-      - uses: actions-rs/cargo@v1
+      - name: Build examples
+        uses: actions-rs/cargo@v1
         with:
           command: build
           args: --examples -v
+      - name: Run example classes
+        uses: actions-rs/cargo@v1
+        with:
+          command: run
+          args: --example classes
 
   doc:
     name: Documentation check

--- a/boa/examples/classes.rs
+++ b/boa/examples/classes.rs
@@ -1,3 +1,5 @@
+// NOTE: this example requires the `console` feature to run correctly.
+
 use boa::{
     class::{Class, ClassBuilder},
     gc::{Finalize, Trace},
@@ -7,11 +9,11 @@ use boa::{
 
 // We create a new struct that is going to represent a person.
 //
-// We derive `Debug`, `Trace` and `Finalize`, It automatically implements `NativeObject`
-// so we can pass it an object in JavaScript.
+// We derive `Debug`, `Trace` and `Finalize`, it automatically implements `NativeObject`
+// so we can pass it as an object in Javascript.
 //
-// The fields of the struct are not accessible by JavaScript unless accessors are created for them.
-/// This  Represents a Person.
+// The fields of the struct are not accessible by Javascript unless we create accessors for them.
+/// Represents a `Person` object.
 #[derive(Debug, Trace, Finalize)]
 struct Person {
     /// The name of the person.
@@ -20,45 +22,46 @@ struct Person {
     age: u32,
 }
 
-// Here we implement a static method for Person that matches the `NativeFunction` signiture.
+// Here we implement a static method for Person that matches the `NativeFunction` signature.
 //
-// NOTE: The function does not have to be implemented of Person it can be a free function,
-// or any function that matches that signature.
+// NOTE: The function does not have to be implemented inside Person, it can be a free function,
+// or any function that matches the required signature.
 impl Person {
-    /// This function says hello
+    /// Says hello if `this` is a `Person`
     fn say_hello(this: &JsValue, _: &[JsValue], context: &mut Context) -> JsResult<JsValue> {
         // We check if this is an object.
         if let Some(object) = this.as_object() {
             // If it is we downcast the type to type `Person`.
             if let Some(person) = object.downcast_ref::<Person>() {
-                // we print the message to stdout.
+                // and print a message to stdout.
                 println!(
                     "Hello my name is {}, I'm {} years old",
                     person.name,
-                    person.age // Here we can access the native rust fields of Person struct.
+                    person.age // Here we can access the native rust fields of the struct.
                 );
                 return Ok(JsValue::undefined());
             }
         }
-        // If `this` was not an object or the type was not an native object `Person`,
+        // If `this` was not an object or the type of `this` was not a native object `Person`,
         // we throw a `TypeError`.
         context.throw_type_error("'this' is not a Person object")
     }
 }
 
 impl Class for Person {
-    // we set the binging name of this function to be `"Person"`.
-    // It does not have to be `"Person"` it can be any string.
+    // We set the binding name of this function to `"Person"`.
+    // It does not have to be `"Person"`, it can be any string.
     const NAME: &'static str = "Person";
     // We set the length to `2` since we accept 2 arguments in the constructor.
     //
     // This is the same as `Object.length`.
-    // NOTE: If this is not defiend that the default is `0`.
+    // NOTE: The default value of `LENGTH` is `0`.
     const LENGTH: usize = 2;
 
-    // This is what is called when we do `new Person()`
+    // This is what is called when we construct a `Person` with the expression `new Person()`.
     fn constructor(_this: &JsValue, args: &[JsValue], context: &mut Context) -> JsResult<Self> {
-        // We get the first argument. If it is unavailable we get `undefined`. And, then call `to_string()`.
+        // We get the first argument. If it is unavailable we default to `undefined`,
+        // and then we call `to_string()`.
         //
         // This is equivalent to `String(arg)`.
         let name = args
@@ -66,12 +69,13 @@ impl Class for Person {
             .cloned()
             .unwrap_or_default()
             .to_string(context)?;
-        // We get the second argument. If it is unavailable we get `undefined`. And, then call `to_u32`.
+        // We get the second argument. If it is unavailable we default to `undefined`,
+        // and then we call `to_u32`.
         //
         // This is equivalent to `arg | 0`.
         let age = args.get(1).cloned().unwrap_or_default().to_u32(context)?;
 
-        // we construct the native struct `Person`
+        // We construct a new native struct `Person`
         let person = Person {
             name: name.to_string(),
             age,
@@ -80,34 +84,35 @@ impl Class for Person {
         Ok(person) // and we return it.
     }
 
-    /// This is where the object is intitialized.
+    /// Here is where the class is initialized.
     fn init(class: &mut ClassBuilder) -> JsResult<()> {
-        // we add a inheritable method `sayHello` with length `0` the amount of args it takes.
+        // We add a inheritable method `sayHello` with `0` arguments of length.
         //
-        // This function is added to `Person.prototype.sayHello()`
+        // This function is added to the `Person` prototype.
         class.method("sayHello", 0, Self::say_hello);
-        // we add a static mathod `is`, and here we use a closure, but it must be convertible
-        // to a NativeFunction. it must not contain state, if it does it will give a compilation error.
+        // We add a static method `is` using a closure, but it must be convertible
+        // to a NativeFunction.
+        // This means it must not contain state, or the code won't compile.
         //
-        // This function is added to `Person.is()`
+        // This function is added to the `Person` class.
         class.static_method("is", 1, |_this, args, _ctx| {
             if let Some(arg) = args.get(0) {
                 if let Some(object) = arg.as_object() {
+                    // We check if the type of `args[0]` is `Person`
                     if object.is::<Person>() {
-                        // we check if the object type is `Person`
-                        return Ok(true.into()); // return `true`.
+                        return Ok(true.into()); // and return `true` if it is.
                     }
                 }
             }
-            Ok(false.into()) // otherwise `false`.
+            Ok(false.into()) // Otherwise we return `false`.
         });
 
-        // Add a inherited property with the value `10`, with default attribute.
-        // (`READONLY, NON_ENUMERABLE, PERMANENT).
+        // We add an `"inheritedProperty"` property to the prototype of `Person` with
+        // a value of `10` and default attribute flags `READONLY`, `NON_ENUMERABLE` and `PERMANENT`.
         class.property("inheritedProperty", 10, Attribute::default());
 
-        // Add a static property with the value `"Im a static property"`, with default attribute.
-        // (`WRITABLE, ENUMERABLE, PERMANENT`).
+        // Finally, we add a `"staticProperty"` property to `Person` with a value
+        // of `"Im a static property"` and attribute flags `WRITABLE`, `ENUMERABLE` and `PERMANENT`.
         class.static_property(
             "staticProperty",
             "Im a static property",
@@ -119,11 +124,14 @@ impl Class for Person {
 }
 
 fn main() {
+    // First we need to create a Javascript context.
     let mut context = Context::new();
 
-    // we register the global class `Person`.
+    // Then we need to register the global class `Person` inside `context`.
     context.register_global_class::<Person>().unwrap();
 
+    // Having done all of that, we can execute Javascript code with `eval`,
+    // and access the `Person` class defined in Rust!
     context
         .eval(
             r"

--- a/boa/src/class.rs
+++ b/boa/src/class.rs
@@ -62,7 +62,7 @@
 
 use crate::{
     builtins::function::NativeFunction,
-    object::{ConstructorBuilder, GcObject, NativeObject, ObjectData},
+    object::{ConstructorBuilder, GcObject, NativeObject, ObjectData, PROTOTYPE},
     property::{Attribute, PropertyDescriptor, PropertyKey},
     Context, JsResult, JsValue,
 };
@@ -87,7 +87,7 @@ pub trait Class: NativeObject + Sized {
 ///
 /// This is automatically implemented, when a type implements `Class`.
 pub trait ClassConstructor: Class {
-    /// The raw constructor that mathces the `NativeFunction` signature.
+    /// The raw constructor that matches the `NativeFunction` signature.
     fn raw_constructor(
         this: &JsValue,
         args: &[JsValue],
@@ -102,9 +102,47 @@ impl<T: Class> ClassConstructor for T {
     where
         Self: Sized,
     {
-        let object_instance = Self::constructor(this, args, context)?;
-        this.set_data(ObjectData::NativeObject(Box::new(object_instance)));
-        Ok(this.clone())
+        if this.is_undefined() {
+            return context.throw_type_error(format!(
+                "cannot call constructor of native class `{}` without new",
+                T::NAME
+            ));
+        }
+
+        let class_constructor =
+            if let Some(obj) = context.global_object().get(T::NAME, context)?.as_object() {
+                obj
+            } else {
+                return context.throw_type_error(format!(
+                    "invalid constructor for native class `{}` ",
+                    T::NAME
+                ));
+            };
+        let class_prototype =
+            if let Some(obj) = class_constructor.get(PROTOTYPE, context)?.as_object() {
+                obj
+            } else {
+                return context.throw_type_error(format!(
+                    "invalid default prototype for native class `{}`",
+                    T::NAME
+                ));
+            };
+
+        let prototype = this
+            .as_object()
+            .and_then(|obj| {
+                obj.get(PROTOTYPE, context)
+                    .map(|o| o.as_object())
+                    .transpose()
+            })
+            .transpose()?
+            .unwrap_or(class_prototype);
+
+        let native_instance = Self::constructor(this, args, context)?;
+        let object_instance = context.construct_object();
+        object_instance.set_prototype_instance(prototype.into());
+        object_instance.borrow_mut().data = ObjectData::NativeObject(Box::new(native_instance));
+        Ok(object_instance.into())
     }
 }
 

